### ISSUE TITLE
Shedding dependencies?

### ISSUE THREAD

**Issue by mindplay-dk** [grain-lang/grain#147]
Interesting language! :-)

Since you don't have a roadmap, I'd like to ask:

* Do you have any plans as far as shedding some of the big dependencies?

* Specifically, do you plan on porting and bootstrapping the compiler?

* Is the dependency on OCaml as intermediary language a permanent feature?

* And are the Node and Yarn dependencies preliminary or permanent?

I really like where this language is going - but it seems like a lot of heavy machinery (two other languages + package managers) for a language that could potentially be very simple, e.g. compiler written in Grain, with a back-end that emits native WASM, and the only compiler + run-time dependency being a WASM/WASI run-time.

Perhaps even better, using something like [Lucet](https://github.com/bytecodealliance/lucet) to build native, stand-alone programs from the WASM output - including the compiler itself.

If this isn't what you have planned, can you say a bit about what you do have planned? 🙂

Thanks!

**Comment by ospencer** [comment]
Hey! Just replied to your comment on #57. 

> Specifically, do you plan on porting and bootstrapping the compiler?

_Eventually._ Right now, our main focus is building out the language itself and making it as easy as possible for contributors to contribute. We're at a decent point with that—we do have fairly heavy dependencies, but everything is handled through yarn. Contributors don't have to figure out how to install OCaml or anything like that. 

There's quite a bit of work that needs to be done with the language before we'd be at a good point to start working on a compiler port to Grain, but we're getting there (we'll need a fully fleshed-out filesystem library, macro system, etc.).

Bootstrapped Grain (let's call it Whole Grain 😂) is a Grain v2.0.0 idea and we're still working towards v0.1.0. Granted, as soon as GC lands, we pretty much have everything we need for v0.1.0.

Once we're at v0.1.0, we'll start distributing binaries, so all of the OCaml and whatnot will be invisible to users. The Node dependency will slowly go away. Node is a dependency largely because portions of our runtime is written is JS, but we're slowly moving all of it to WebAssembly. Ideally Grain will be able to be run on any WASI-compliant runtime.

Long story short, that's a potential plan for the future, but we've got quite a few hurdles to clear before we get there. Right now we're continuing to put all of our effort towards core language features.

**Comment by mindplay-dk** [comment]
Very good answers, thanks! 🙂👍

One more question, if you don't mind: what's the story with regards to parallelism?

Threads, fibers, workers, coroutines... something else?

**Comment by ospencer** [comment]
Of course. We haven't talked about it much, but my guess right now would be threads since [WebAssembly should make that pretty easy](https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md). 👍

**Comment by mindplay-dk** [comment]
And what would thread support look like in the language?

Async/await maybe? Or just an API with threading primitives?

I suppose locks and mutexes wouldn't be required since everything is immutable. (except boxes?)

Some kind of lightweight threading primitive like fibers might be worth considering though - proper thread scheduling is hard, and most modern languages offer something a little more abstract, so that things like web/socket servers can be built to scale safely and easily. (?)

**Comment by ospencer** [comment]
I can't yet say what thread support will look like yet, since I haven't put much thought towards it. Since we're aiming to make things like this as easy as possible for users, I imagine that we'll have a decent level of abstraction with whatever we do. Fibers are something to consider, though if we do have them we'll probably still also have access to threads.

Generally, you're right—most kinds of computation won't need locks/mutexes which should just in general make parallelism fairly easy, though we'll have to keep in mind that we'll need to make it just as easy for mutable structures as well (to adhere to our language goals).

**Comment by mindplay-dk** [comment]
Yeah, threads would be useful as well and might be an easier place to start. Makes sense.

Well, thanks for your time! I will definitely be watching this project to see where it goes. 🙂✌️

**Comment by ospencer** [comment]
Of course! :D
